Merge develop into main (#274)
DataDog/dd-sdk-reactnative: docs/codepush.md

# CodePush for React Native support

## Overview

Enable React Native Crash Reporting and Error Tracking to get comprehensive crash reports and error trends with Real User Monitoring.

Each time you release a new [CodePush][1] version for your React Native application, you need to upload the source maps to Datadog to unminify errors.

> To achieve this, we recommend using `@datadog/mobile-react-native-code-push` in your app and the [datadog-ci][3] `react-native codepush` command to upload your source maps. It will ensure that the version will be consistent in both reported crashes and uploaded source maps.

Reviewer: ```suggestion
Datadog recommends using `@datadog/mobile-react-native-code-push` in your app and the [datadog-ci][3] `react-native codepush` command to upload your source maps. This ensures that the `version` is consistent in both reported crashes and uploaded source maps.
```

## Setup

Start by following the [installation steps][2] for installing `@datadog/mobile-react-native`.

Install `@datadog/mobile-react-native-code-push`.

To install with NPM, run:

```sh
npm install @datadog/mobile-react-native-code-push
```

To install with Yarn, run:

```sh
yarn add @datadog/mobile-react-native-code-push
```

Replace `DdSdkReactNative.initialize` by `DatadogCodepush.initialize` in your code:

```js
import { DdSdkReactNativeConfiguration } from '@datadog/mobile-react-native';
import { DatadogCodepush } from '@datadog/mobile-react-native-code-push';

const config = new DdSdkReactNativeConfiguration(
    '<CLIENT_TOKEN>',
    '<ENVIRONMENT_NAME>',
    '<RUM_APPLICATION_ID>',
    true, // track User interactions (e.g.: Tap on buttons. You can use 'accessibilityLabel' element property to give tap action the name, otherwise element type will be reported)
    true, // track XHR Resources

>     true // track Errors

Reviewer 2: this (and the suggestion above) is an example of something that really might be like this in the code, as in the comments might be capitalized there too

);

await DatadogCodepush.initialize(config);
```

## Upload CodePush source maps

Install [`@datadog/datadog-ci`][3] as a development dependency to your project.

To install it with npm:

```sh
npm install @datadog/datadog-ci --save-dev
```

To install it with yarn:

```sh
yarn add -D @datadog/datadog-ci
```

> Create a gitignored `datadog-ci.json` file at the root of your project containing your API key and the Datadog site (if not `datadoghq.com`):

Reviewer: ```suggestion
Create a gitignore `datadog-ci.json` file at the root of your project containing your API key and the Datadog site (if not `datadoghq.com`):
```

```json
{
    "apiKey": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    "site": "datadoghq.eu"
}
```

> **Note:** You can also export them as `DATADOG_API_KEY` and `DATADOG_SITE` environment variables.

Reviewer: ```suggestion
You can also export them as `DATADOG_API_KEY` and `DATADOG_SITE` environment variables.
```

> When releasing a new CodePush bundle, specify a directory for outputting the source maps and bundle:

Reviewer: ```suggestion
When releasing a new CodePush bundle, specify a directory to output the source maps and bundle:
```

```sh
appcenter codepush release-react -a MyOrganization/MyApplication -d MyDeployment --sourcemap-output --output-dir ./build
```

> Then run the `datadog-ci react-native codepush` command by passing the adequate CodePush `app` and `deployment` arguments.

Reviewer: ```suggestion
Run the `datadog-ci react-native codepush` command by passing the adequate CodePush `app` and `deployment` arguments.
```

To run it with npm:

```sh
npm run datadog-ci react-native codepush --platform ios --service com.company.app --bundle ./build/CodePush/main.jsbundle --sourcemap ./build/CodePush/main.jsbundle.map --app MyOrganization/MyApplication --deployment MyDeployment
```

> To run it with yarn:

Reviewer: ```suggestion
To run it with Yarn:
```

```sh
yarn datadog-ci react-native codepush --platform ios --service com.company.app --bundle ./build/CodePush/main.jsbundle --sourcemap ./build/CodePush/main.jsbundle.map --app MyOrganization/MyApplication --deployment MyDeployment
```

## Alternatives

> These steps will ensure that the `version` will match the format `{commercialVersion}-codepush.{codePushLabel}`, such as `1.2.4-codepush.v3`.

Reviewer: ```suggestion
These steps ensure that the `version` matches the format `{commercialVersion}-codepush.{codePushLabel}`, such as `1.2.4-codepush.v3`.
```

You can also do that by specifying a `versionSuffix` in the SDK configuration:

```js
const config = new DdSdkReactNativeConfiguration(
    '<CLIENT_TOKEN>',
    '<ENVIRONMENT_NAME>',
    '<RUM_APPLICATION_ID>',
    true, // track User interactions (e.g.: Tap on buttons. You can use 'accessibilityLabel' element property to give tap action the name, otherwise element type will be reported)
    true, // track XHR Resources
    true // track Errors
);

config.versionSuffix = `codepush.${codepushVersion}`; // will result in "1.0.0-codepush.v2"
```

> Be aware that to avoid potential version clashes, `versionSuffix` adds a dash (`-`) before the suffix.
> You can obtain the `codepushVersion` by hardcoding it or using [`CodePush.getUpdateMetadata`][4].

Reviewer: ```suggestion
In order to avoid potential version clashes, the `versionSuffix` adds a dash (`-`) before the suffix.

To obtain the `codepushVersion`, you can hardcode it or use [`CodePush.getUpdateMetadata`][4].
```

> You can upload your source maps using the [`datadog-ci react-native upload`][5] command, making sure the `--release-version` argument matches the one set in the SDK configuration.

Reviewer: ```suggestion
Then, upload your source maps using the [`datadog-ci react-native upload`][5] command, and ensure the `--release-version` argument matches the one set in the SDK configuration.
```

```sh
appcenter codepush release-react -a MyOrganization/MyApplication -d MyDeployment --sourcemap-output --output-dir ./build
```

Run the `datadog-ci react-native codepush` command by passing the adequate CodePush `app` and `deployment` arguments.

To run it with NPM:

```sh
npm run datadog-ci react-native codepush --platform ios --service com.company.app --bundle ./build/CodePush/main.jsbundle --sourcemap ./build/CodePush/main.jsbundle.map --app MyOrganization/MyApplication --deployment MyDeployment
```

To run it with Yarn:

```sh
yarn datadog-ci react-native codepush --platform ios --service com.company.app --bundle ./build/CodePush/main.jsbundle --sourcemap ./build/CodePush/main.jsbundle.map --app MyOrganization/MyApplication --deployment MyDeployment
```

## Alternatives

These steps ensure that the `version` matches the format `{commercialVersion}-codepush.{codePushLabel}`, such as `1.2.4-codepush.v3`.

You can also do that by specifying a `versionSuffix` in the SDK configuration:

```js
const config = new DdSdkReactNativeConfiguration(
    '<CLIENT_TOKEN>',
    '<ENVIRONMENT_NAME>',
    '<RUM_APPLICATION_ID>',

>     true, // track User interactions (e.g.: Tap on buttons. You can use 'accessibilityLabel' element property to give tap action the name, otherwise element type will be reported)
>     true, // track XHR Resources
>     true // track Errors
> );
> 
> config.versionSuffix = `codepush.${codepushVersion}`; // will result in "1.0.0-codepush.v2"

Reviewer: ```suggestion
    true, // track user interactions (such as a tap on buttons). You can use the 'accessibilityLabel' element property to give the tap action a name, otherwise the element type is reported
    true, // track XHR resources
    true // track errors
);

config.versionSuffix = `codepush.${codepushVersion}`; // Results in "1.0.0-codepush.v2"
```
